### PR TITLE
[REVIEW] Add Java count aggregation to include null values [skip ci]

### PR DESCRIPTION
**Pull request by nartal1**
This PR adds support for count aggregation to include null values. COUNT_VALID counts only valid values excluding null values. COUNT_ALL includes null values

### CODE REVIEW

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 1]
```diff
@@ -985,7 +985,12 @@ public static OrderByArg desc(final int index, final boolean isNullSmallest) {
   }
 
   public static Aggregate count(int index) {
-    return Aggregate.count(index);
+    return Aggregate.count(index, false);
+  }
+
+  public static Aggregate count(int index, boolean include_nulls) {
+    assert include_nulls == true || include_nulls == false : "include_nulls must be boolean";
```
Comment (razajafri) [inline java/src/main/java/ai/rapids/cudf/Table.java]: unnecessary assert

Comment (nartal1) [reply java/src/main/java/ai/rapids/cudf/Table.java]: Thanks for reviewing. Removed it.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 2]
```diff
@@ -1104,7 +1104,7 @@ void testWindowStatic() {
       }
 
       try (ColumnVector expected = ColumnVector.fromInts(2, 3, 3, 3, 2);
```
Comment (razajafri) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: sorry, but can you include nulls in this vector to really know if its counting valids only? 

can you also add a test for count_all?

File: java/src/test/java/ai/rapids/cudf/TableTest.java [hunk 3]
```diff
@@ -1653,6 +1653,42 @@ void testGroupByCountWithNulls() {
...
+        assertEquals(2, t3.getRowCount());
+
+        // compare the grouping columns
+        assertTrue(groupCol.isNull(0));
+        assertEquals(groupCol.getInt(1), 1);
+
+        // compare the agg columns
+        // count(1)
+        assertEquals(countCol.getInt(0), 2);
+        assertEquals(countCol.getInt(1), 4);
```
Comment (abellina) [inline java/src/test/java/ai/rapids/cudf/TableTest.java]: I think this one also includes nulls? Would be good to add a count somewhere (not of the first col obviously) that doesn't count nulls in this test, for completeness.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java [hunk 4]
```diff
@@ -1104,7 +1104,7 @@ void testWindowStatic() {
...
-           ColumnVector result = v1.rollingWindow(AggregateOp.COUNT, options)) {
+           ColumnVector result = v1.rollingWindow(AggregateOp.COUNT_VALID, options)) {
```
Comment (abellina) [inline java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: Pinging @mythrocks for visibility. I do wonder if we want to wrap the ops for rolling window, if that makes sense... so we don't depend on the enum outside of jni cudf? That said @nartal1, I don't think that's a change for this PR.

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 5]
```diff
@@ -985,7 +985,11 @@ public static OrderByArg desc(final int index, final boolean isNullSmallest) {
   }
 
   public static Aggregate count(int index) {
-    return Aggregate.count(index);
+    return Aggregate.count(index, false);
```
Comment (abellina) [inline java/src/main/java/ai/rapids/cudf/Table.java]: I agree this is probably the choice that makes the most sense for `count(index)`.

Could you add function headers? This is not directly related to your change, but it would be nice to document what `count(index)` does in a header clearly.

File: java/src/main/java/ai/rapids/cudf/Aggregate.java [hunk 6]
```diff
@@ -30,10 +30,13 @@ private Aggregate(AggregateOp type, int index) {
     this.index = index;
   }
 
-  static Aggregate count(int index) {
-    return new Aggregate(AggregateOp.COUNT, index);
+  // Include null in count if include_nulls = true
+  static Aggregate count(int index, boolean include_nulls) {
+    if (include_nulls == true) {
+      return new Aggregate(AggregateOp.COUNT_ALL, index);
```
Comment (abellina) [inline java/src/main/java/ai/rapids/cudf/Aggregate.java]: could do
```
return new Aggregate(include_nulls ? 
  AggregateOp.COUNT_ALL : 
  AggregateOp.COUNT_VALID,
  index)
```

But I would at least loose the `== true`.

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java (hunk 4 above)
Comment (nartal1) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: I had to update it since we don't have `COUNT` anymore.  Based on @razajafri 's comment I was planning to add a new test something like below. Please let me know how to go about this.
`
 @Test
  void testWindowStaticCounts() {
    WindowOptions options = WindowOptions.builder().window(2, 1)
            .minPeriods(2).build();
    try (ColumnVector v1 = ColumnVector.fromBoxedInts(5, 4, null, 6, 8)) {
      try (ColumnVector expected = ColumnVector.fromInts(2, 2, 2, 2, 2);
           ColumnVector result = v1.rollingWindow(AggregateOp.COUNT_VALID, options)) {
        assertColumnsAreEqual(expected, result);
      }
      try (ColumnVector expected = ColumnVector.fromInts(2, 3, 3, 3, 2);
           ColumnVector result = v1.rollingWindow(AggregateOp.COUNT_ALL, options)) {
        assertColumnsAreEqual(expected, result);
      }
    }
  }`

Comment (mythrocks) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: You mean like a separate `WindowAggregateOp` type? That would still depend on `cudf::experimental::aggregate::Kind`, at the moment. :/

**Comment by nartal1** [comment]
@razajafri @abellina @mythrocks  PTAL. Please let me know if the test in ColumnVector.java is acceptable OR should i just keep the old one and update COUNT to COUNT_VALID.

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 5 above)
Comment (abellina) [reply java/src/main/java/ai/rapids/cudf/Table.java]: @nartal1 mind adding the function headers for the other aggregate ops while you are in there?

Comment (nartal1) [reply java/src/main/java/ai/rapids/cudf/Table.java]: Added function headers for other aggregate ops

File: java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java (hunk 4 above)
Comment (abellina) [reply java/src/test/java/ai/rapids/cudf/ColumnVectorTest.java]: @mythrocks ok if it is not going to lend itself to a similar interface from the group by aggregate, that is fine. I was mostly curious.

File: java/src/main/java/ai/rapids/cudf/Table.java (hunk 5 above)
Comment (abellina) [reply java/src/main/java/ai/rapids/cudf/Table.java]: @nartal1 thanks for making changes, but could you add that the other functions are all skipping nulls? That was the part I was most interested in.

Comment (nartal1) [reply java/src/main/java/ai/rapids/cudf/Table.java]: Thanks @abellina . Updated it.

**Comment by nartal1** [comment]
Thanks @abellina .  @razajafri, could you PTAL.
@mythrocks FYI, minor change in staticwidow test.

**Comment by revans2** [comment]
@kkraus14 could we get this merged in?
